crates/re_ui/src/list_item2/scope.rs

```diff
@@ -0,0 +1,137 @@
+                    Wrap in a `list_item_scope`."
+                );
+            }
+            state.cloned().unwrap_or_default()
+        })
+    }
+
+    fn peek(ctx: &egui::Context) -> Option<&State> {
+        ctx.data(|reader| reader.get_temp(*STATE_STACK_ID))
+    }
+}
+
+/// Create a scope in which `[ListItem]`s can be created.
+///
+/// This scope serves two purposes:
+/// - Manage the state that is saved across frame (e.g. for tracking column boundary position).
+/// - Manage the range of X coordinates defining the boundaries of the hover/selection highlight.
+///
+/// State is loaded against the provided `id`, and pushed to a global stack, such that calls to this
+/// function may be nested. `ListItem` code will always use the top of the stack as current state.
+///
+/// The hover/selection highlight coordinate range is determined with the following heuristics:
+/// - Value passed as argument if not `None`.
+/// - Value from the parent scope if the scope is nested.
+/// - Clip rectangle's `x_range` (legacy behavior).
+///
+/// Given the above, `list_item_scope` can be used for two potentially distinct use-cases:
+/// 1) Store a suitable `background_x_range` value for use by nested `ListItem`s. This happens,
+///    e.g., close to the top of the `egui::SidePanel::show` closure, where the panel size
+///    information is readily available (e.g. `ui.max_rect().x_range()`).
+/// 2) Limit state sharing for a subgroup of `ListItem`s. This makes it possible to independently
+///    align the columns of two `ListItem`s subgroups, for which a single, global alignment would
+///    be detrimental. This may happen in deeply nested UI code.
+///
+/// TODO(#6156): the background X range stuff is to be split off and generalised for all full-span
+/// widgets.
+pub fn list_item_scope<R>(
+    ui: &mut egui::Ui,
+    id: impl Into<egui::Id>,
+    background_x_range: Option<egui::Rangef>,
+    content: impl FnOnce(&mut egui::Ui) -> R,
+) -> R {
+    /*
+    data contains two set of things:
+    - some per container state
+    - a global state stack that is read by actual list items
+     */
+
+    let id = id.into();
+
+    // read the state for this container, if any
+    let state: Option<State> = ui.data(|reader| reader.get_temp(id));
+    let mut state = state.unwrap_or_default();
+
+    // determine the background x range to use
+    state.background_x_range = if let Some(background_x_range) = background_x_range {
+        background_x_range
+    } else if let Some(parent_state) = StateStack::peek(ui.ctx()) {
+        parent_state.background_x_range
+    } else {
+        ui.clip_rect().x_range()
+    };
+
+    // push, run, pop
+    StateStack::push(ui.ctx(), state.clone());
+    let result = content(ui);
+    let state = StateStack::pop(ui.ctx());
+
+    // save the state for this container
+    if let Some(state) = state {
+        ui.data_mut(|writer| writer.insert_temp(id, state));
+    }
```

Comment:
I'll just ignore all of this for this PR, since you say it's changing later anyways